src/NHibernate.Test/NHSpecificTest/GH0831/FixtureByCode.cs

```diff
@@ -0,0 +1,237 @@
+			});
+		}
+
+		[Test]
+		public void CanHandleMultiply()
+		{
+			Assert.Multiple(() =>
+			{
+				CanFilter(e => decimal.Multiply(e.EntityValue, 10m) > 10m);
+				CanFilter(e => decimal.Multiply(10m, e.EntityValue) > 10m);
+
+				CanSelect(e => decimal.Multiply(e.EntityValue, 10m));
+				CanSelect(e => decimal.Multiply(10m, e.EntityValue));
+			});
+		}
+
+		[Test]
+		public void CanHandleNegate()
+		{
+			Assert.Multiple(() =>
+			{
+				CanFilter(e => decimal.Negate(e.EntityValue) > -1.0m);
+				CanSelect(e => decimal.Negate(e.EntityValue));
+			});
+		}
+
+		[Test]
+		public void CanHandleRemainder()
+		{
+			Assume.That(TestDialect.SupportsModuloOnDecimal, Is.True);
+
+			Assert.Multiple(() =>
+			{
+				CanFilter(e => decimal.Remainder(e.EntityValue, 2m) == 0);
+				CanFilter(e => decimal.Remainder(2m, e.EntityValue) < 1);
+
+				CanSelect(e => decimal.Remainder(e.EntityValue, 2m));
+				CanSelect(e => decimal.Remainder(2m, e.EntityValue));
+			});
+		}
+
+		[Test]
+		public void CanHandleRound()
+		{
+			AssumeFunctionSupported("round");
+
+			Assert.Multiple(() =>
+			{
+				CanFilter(e => decimal.Round(e.EntityValue) >= 2.0m);
+				CanFilter(e => decimal.Round(e.EntityValue, 1) >= 1.5m);
+
+				// SQL round() always rounds up. 
+				CanSelect(e => decimal.Round(e.EntityValue), entities.Select(e => decimal.Round(e.EntityValue, MidpointRounding.AwayFromZero)));
+				CanSelect(e => decimal.Round(e.EntityValue, 1), entities.Select(e => decimal.Round(e.EntityValue, 1, MidpointRounding.AwayFromZero)));
+			});
+		}
+
+		[Test]
+		public void CanHandleSubtract()
+		{
+			Assert.Multiple(() =>
+			{
+				CanFilter(e => decimal.Subtract(e.EntityValue, 1m) > 1m);
+				CanFilter(e => decimal.Subtract(2m, e.EntityValue) > 1m);
+
+				CanSelect(e => decimal.Subtract(e.EntityValue, 1m));
+				CanSelect(e => decimal.Subtract(2m, e.EntityValue));
+			});
+		}
+
+		private void CanFilter(Expression<Func<Entity, bool>> predicate)
+		{
+			using (ISession session = OpenSession())
+			using (session.BeginTransaction())
+			{
+				IEnumerable<Entity> inMemory = entities.Where(predicate.Compile()).ToList();
+				IEnumerable<Entity> inSession = session.Query<Entity>().Where(predicate).ToList();
+
+				CollectionAssert.AreEquivalent(inMemory, inSession);
+			}
+		}
+
+		private void CanSelect(Expression<Func<Entity, decimal>> predicate)
+		{
+			IEnumerable<decimal> inMemory = entities.Select(predicate.Compile()).ToList();
+
+			CanSelect(predicate, inMemory);
+		}
+
+		private void CanSelect(Expression<Func<Entity, decimal>> predicate, IEnumerable<decimal> expected)
+		{
+			using (ISession session = OpenSession())
+			using (session.BeginTransaction())
+			{
+				IEnumerable<decimal> inSession = null;
+				Assert.That(() => inSession = session.Query<Entity>().OrderBy(e => e.EntityValue).Select(predicate).ToList(), Throws.Nothing);
+
+				Assert.That(inSession, Is.EqualTo(expected.OrderBy(x => x)).Within(0.001M));
```

Comment:
The order by of `expected.OrderBy(x => x)` must be removed. It is ordering by computed value instead of preserving the ordering by original value the in memory source has.